kinto/core/resource/__init__.py

```diff
@@ -394,20 +431,45 @@ def _plural_get(self, head_request):
             include_deleted=include_deleted,
         )
 
-        offset = offset + len(objects)
+        current_offset = self._extract_offset() if "_offset" in self.request.validated["querystring"] else 0
+        used_offset_pagination = (pagination_rules == [] and "_offset" in self.request.validated["querystring"])
 
-        if limit and len(objects) == limit + 1:
-            lastobject = objects[-2]
-            next_page = self._next_page_url(sorting, limit, lastobject, offset)
-            headers["Next-Page"] = next_page
+        if used_offset_pagination and current_offset > 0:
+            # Apply manual offset slicing
+            objects = objects[current_offset:]
+        offset = offset + len(objects)
+        if used_offset_pagination:
+            if limit and len(objects) == limit + 1:
+                next_url = self._build_offset_next_url(limit, current_offset)
+        else:
+            if limit and len(objects) == limit + 1:
+                    lastobject = objects[-2]
+                    next_page = self._next_page_url(sorting, limit, lastobject, offset)
+                    headers["Next-Page"] = next_page
 
         if partial_fields:
             objects = [dict_subset(obj, partial_fields) for obj in objects]
 
         # See bigger explanation above about the use of limits. The need for slicing
         # here is because we might have asked for 1 more object just to see if there's
         # a next page. But we have to honor the limit in our returned response.
-        return self.postprocess(objects[:limit])
+        result_objects = objects[:limit]
+        if used_offset_pagination:
+            response_data = {
+                'data': result_objects,
+                'pagination': {
```

Comment:
I'm not sure about this new `pagination` field. If we add it, we must check that it is consistent with everything else, and well documented. There is also the option to use response headers. Kinto responses haven't changed in 10 years 😅